### PR TITLE
SNMP schema proposal

### PR DESCRIPTION
**Pull request by tsvanduyn**
This change is a proposal to move SNMP to the configdb.

### CODE REVIEW

File: doc/snmp-schema-addition.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# SONiC SNMP TABLE schema proposal #
+
+## Scope of the change ##
+
+Currently SNMP configuration is managed from a mix of yaml files and DB based ACLs, we propose to integrate the SMP configs into an SNMP table in the DB.
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:5]: >SMP configs [](start = 112, length = 11)

Do you mean "SNMP yaml config file" ?

File: doc/snmp-schema-addition.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+In the end this is used to produce */etc/snmp/snmpd.conf*.
+
+## Limits and incentives to change ##
+- This goes against the principle of having the config DB be the central repository of configuration data.
+- Imposes the provisioning of multiple static files.
+- the snmpContact is hardcoded in the */usr/share/sonic/templates/snmpd.conf.j2* file.
+
+## Proposal: integrate the SNMP configuration into config DB ##
+### Proposed Schema ###
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:26]: >Schema [](start = 13, length = 6)

Please reformat your json. There are many online/offline formatter.

File: doc/snmp-schema-addition.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+```
+{"SNMP": [ {"LOCATION": "LOCATION_STRING",
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:29]: >[ [](start = 9, length = 1)

why we need this array?

File: doc/snmp-schema-addition.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            "CONTACT": "CONTACT_STRING",
+            "SNMP_COMMUNITIES": [ {"SNMP_COMMUNITY": "SNMP_COMMUNITY_STRING1", "TYPE":"RO|RW"},
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:31]: >SNMP_COMMUNITIES [](start = 13, length = 16)

Currently our ConfigDB has these levels in any database: Table / Key / Field / Value. So we could not implement too many levels in json.
ref: [https://github.com/Azure/SONiC/blob/2a3be770db988c3d089ac54d6ee3eb9bb77ae621/doc/config_db.json](https://github.com/Azure/SONiC/blob/2a3be770db988c3d089ac54d6ee3eb9bb77ae621/doc/config_db.json)
All the above level/below levels are key-value pairs, so we don't use json array.
Suggest you treat SNMP_COMMUNITY_STRING1 as key, and other field/values pairs as its value.

File: doc/snmp-schema-addition.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                                  {"SNMP_COMMUNITY": "SNMP_COMMUNITY_STRING2", "TYPE":"RO|RW"},
+                                  {"SNMP_COMMUNITY": "SNMP_COMMUNITY_STRING3", "TYPE":"RO|RW"}],
+            "SNMP_USERS": [ {"SNMP_USER": "SNMP_USER1_STRING",
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:34]: >[ [](start = 26, length = 1)

Do not use array. treat SNMP_USER1_STRING as a key.

File: doc/snmp-schema-addition.md (hunk 1 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md:5]: Yes I mean the snmp.yml file.

File: doc/snmp-schema-addition.md (hunk 3 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md:29]: This is standard on every networking vendor I've ever worked with.   I thought that there would be other companies out there that might use this.

File: doc/snmp-schema-addition.md (hunk 2 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md:26]: Done.

File: doc/snmp-schema-addition.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+   "SNMP":{
+      "LOCATION":{
+         "SNMP_LOCATION":"<SNMP_LOCATION_STRING>"
+      },
+      "CONTACT":{
+         "SNMP_CONTACT":"<SNMP_CONTACT_STRING>"
+      }
+   }
+}{
+   "SNMP_COMMUNITIES":{
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:39]: Should table name be singular, e.g., `SNMP_COMMUNITY`?

File: doc/snmp-schema-addition.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+      "<SNMP_COMM>":{
+         "Type":"RO|RW"
+      },
+      "<SNMP_COMM>":{
+         "Type":"RO|RW"
+      },
+      "<SNMP_COMM>":{
+         "Type":"RO|RW"
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:47]: `Type` here is not all-caps and doesn't align with other key names. Is this intentional?

File: doc/snmp-schema-addition.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+- SNMP_USER_AUTH_PASSWORD: String, defines which authentication password will be used for that SNMP user.
+- SNMP_USER_ENCRYPTION_PASSWORD: String, defines which encryption password will be used for that SNMP user. 
+
+
+
+New keys:
+- "type":  Optional, if ommited defaults to 'ro', there are 2 possible values:  
+           "ro": read-only, the only implemented method at this time.  
+           "rw": well you never know - here for completeness but unused in the code.  
+
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:98]: Cases of "type", "ro" and "rw" do not match the above schema. Please unify case throughout the document.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: doc/snmp-schema-addition.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+{
+   "SNMP":{
+      "LOCATION":{
+         "SNMP_LOCATION":"<SNMP_LOCATION_STRING>"
+      },
+      "CONTACT":{
+         "SNMP_CONTACT":"<SNMP_CONTACT_STRING>"
+      }
+   }
+}{
```
Comment (qiluo-msft) [inline doc/snmp-schema-addition.md:38]: Seems invalid json syntax

File: doc/snmp-schema-addition.md (hunk 6 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md]: Updated to be SNMP_COMMUNITY

File: doc/snmp-schema-addition.md (hunk 7 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md]: Updated Type to be TYPE

File: doc/snmp-schema-addition.md (hunk 8 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md]: Fixed.

File: doc/snmp-schema-addition.md (hunk 9 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md]: Updated syntax.  From this page: 
https://jsonformatter.curiousconcept.com/# 
I tested each of the three sections.  If there is something else you want can you please be a bit more specific? 

Thanks.

File: doc/snmp-schema-addition.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+Currently SNMP configuration is managed from a mix of yaml files and DB based ACLs, we propose to integrate the SNMP configs into an SNMP table in the DB.  
+The following document proposes the database Schema in Json format and the list of required changes to the code.
+
+## Current configurtions for SNMP ##
+sonic-buidimage and sonic-snmpagent depend on the following files:
+1. */etc/sonic/snmp.yaml*
+    defines 2 keys:  
+        *snmp_rocommunity:*  
+        *snmp_location:*
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:13]: These are not indented in the rendered Markdown. Suggest making them bullet points.

File: doc/snmp-schema-addition.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+- SNMP_USER_ENCRYPTION_TYPE: String, defines which encryption type will be used for that SNMP user DES or AES. 
+- SNMP_USER_AUTH_PASSWORD: String, defines which authentication password will be used for that SNMP user.
+- SNMP_USER_ENCRYPTION_PASSWORD: String, defines which encryption password will be used for that SNMP user. 
+
+
+
+New keys:
+- "TYPE":  Optional, if ommited defaults to 'ro', there are 2 possible values:  
+           "RO": read-only, the only implemented method at this time.  
+           "RW": well you never know - here for completeness but unused in the code.
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:99]: These are not indented in the rendered Markdown. Suggest making them bullet points.

File: doc/snmp-schema-addition.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+The changes we propose are only additive to remain compatible with the current install base and the current way of doing things.
+
+In repo *sonic-buildimage*:
+
+*dockers/docker-snmp-v2/snmpd.conf.j2*:  
+    verify the existence of the SNMP table in the datatbase and fork behavior if present, if not continue using old method.
+
+*dockers/docker-snmp-v2/snmpd-config-updater*:  
+    this file will be deprecated soon by caclmgrd so no updates will be done
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:111]: These should be bullet points so they are indented under `In repo *sonic-buildimage*:` in the rendered Markdown.

File: doc/snmp-schema-addition.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+
+In repo *sonic-swss-common*: 
+
+*common/schema.h*:  
+```
+#define CFG_SNMP_TABLE_NAME           "SNMP"
+```
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:119]: This should be indented under `In repo *sonic-swss-common*:` in the rendered Markdown.

File: doc/snmp-schema-addition.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+In repo *sonic-swss*:
+
+*doc/swss-schema.md*:
+    add the definition of this schema
```
Comment (jleveque) [inline doc/snmp-schema-addition.md:124]: This should be indented under `In repo *sonic-swss-common*:` in the rendered Markdown.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: doc/snmp-schema-addition.md (hunk 9 above)
Comment (qiluo-msft) [reply doc/snmp-schema-addition.md]: Still invalid json `{ .. } { ... }`

---
In reply to: [527913475](https://github.com/Azure/SONiC/pull/696#discussion_r527913475) [](ancestors = 527913475)

**Review by qiluo-msft — APPROVED** [review]

File: doc/snmp-schema-addition.md (hunk 9 above)
Comment (tsvanduyn) [reply doc/snmp-schema-addition.md]: This is actually 3x different JSON sections to match our redis DB and what we already have for snmptraps.  I've updated the PR to make that more clear.

**Review by jleveque — APPROVED** [review]

**Comment by SuvarnaMeenakshi** [comment]
Can you move this doc to doc/snmp directory?

**Review by SuvarnaMeenakshi — APPROVED** [review]

**Pull request merged**